### PR TITLE
SW-964 App and Email notifications for Accession State Summaries

### PR DESCRIPTION
**Pull request by karthikbtf**
Allow user to receive in-app and email notifications for existing accession state summaries, generated by the daily task job.

### CODE REVIEW

File: src/main/kotlin/com/terraformation/backend/email/EmailNotificationService.kt [hunk 1]
```diff
@@ -101,8 +112,7 @@ class EmailNotificationService(
     emailService.sendUserNotification(
         user,
         "userAddedToOrganization",
-        UserAddedToOrganization(config, admin, organization, organizationHomeUrl),
-        requireOptIn = false)
+        UserAddedToOrganization(config, admin, organization, organizationHomeUrl))
```
Comment (karthikbtf) [inline src/main/kotlin/com/terraformation/backend/email/EmailNotificationService.kt]: This one change is unrelated.. should we skip sending this email if user opted out of email comms?

File: src/main/kotlin/com/terraformation/backend/customer/AppNotificationService.kt [hunk 2]
```diff
@@ -165,19 +169,59 @@ class AppNotificationService(
...
+      notificationType: NotificationType,
+      message: NotificationMessage,
+      localUrl: URI
+  ) {
     val projectId = parentStore.getProjectId(facilityId)!!
     val organizationId = parentStore.getOrganizationId(projectId)!!
-    val emailR = projectStore.fetchEmailRecipients(projectId, false)
-    if (emailR.isEmpty()) {
-      return
-    }
```
Comment (karthikbtf) [inline src/main/kotlin/com/terraformation/backend/customer/AppNotificationService.kt:180]: debug code that got checked in earlier

File: src/main/resources/templates/email/accessionsAwaitingProcessing/body.ftlh.mjml [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+<mjml>
+    <mj-include path="../head.ftlh.mjml" />
+
+    <mj-body>
+        <mj-include path="../logo.ftlh.mjml" />
+
+        <mj-section padding="0px">
+            <mj-column mj-class="body-wrapper">
+                <mj-text mj-class="text-headline06-bold">
+                    <!-- htmlmin:ignore --><#if numAccessions == 1><!-- htmlmin:ignore -->
```
Comment (karthikbtf) [inline src/main/resources/templates/email/accessionsAwaitingProcessing/body.ftlh.mjml:11]: The  `<!-- htmlmin:ignore -->` padding is needed to enable FreeMarker conditional templating.
See https://documentation.mjml.io/#mj-raw

Comment (karthikbtf) [reply src/main/resources/templates/email/accessionsAwaitingProcessing/body.ftlh.mjml:11]: All emails were tested, with count=1 and count != 1

File: src/main/kotlin/com/terraformation/backend/email/WebAppUrls.kt [hunk 4]
```diff
@@ -64,6 +65,18 @@ class WebAppUrls(private val config: TerrawareServerConfig) {
     return UriBuilder.fromPath(accessionGerminationTestPath(accessionId, testType)).build()
   }
 
+  fun fullAccessions(organizationId: OrganizationId, state: AccessionState): URI {
+    return UriBuilder.fromUri(config.webAppUrl)
+        .path("/accessions")
+        .queryParam("stage", state)
```
Comment (karthikbtf) [inline src/main/kotlin/com/terraformation/backend/email/WebAppUrls.kt:76]: stage=<state> param is for UI to show corresponding views in the table. To be implemented in the frontend by me shortly (currently unsupported).

File: src/main/kotlin/com/terraformation/backend/email/EmailNotificationService.kt (hunk 1 above)
Comment (sgrimm) [reply src/main/kotlin/com/terraformation/backend/email/EmailNotificationService.kt]: No, this one is supposed to be sent whether or not the user opted out (in fact, it's the reason there's a `requireOptIn` parameter at all!)

Comment (sgrimm) [reply src/main/kotlin/com/terraformation/backend/email/EmailNotificationService.kt]: One reason is that adding a nonexistent user to an organization causes their account to be created, and they won't have been able to opt in at that point.

File: src/main/kotlin/com/terraformation/backend/seedbank/daily/StateSummaryNotificationTask.kt [hunk 5]
```diff
@@ -128,13 +133,30 @@ class StateSummaryNotificationTask(
...
+              eventPublisher.publishEvent(
+                  AccessionsReadyForTestingEvent(facilityId, count, weeks, state))
+            }
+          }
+          AccessionState.Dried ->
+              eventPublisher.publishEvent(AccessionsFinishedDryingEvent(facilityId, count, state))
+          else -> log.warn("Unsupported state $state for notification events.")
+        }
+      } catch (e: Exception) {
+        log.error("Error handling notification event for state summary $state", e)
```
Comment (sgrimm) [inline src/main/kotlin/com/terraformation/backend/seedbank/daily/StateSummaryNotificationTask.kt]: The fact that this is sending email makes it less obvious what the right behavior is here, but this is going to be a behavior difference from the previous version of the code if there's an exception while publishing the notifications.

Previously, the exception would have bubbled up to `DailyTaskRunner.runTask()`, which would cause it to not mark the current day as successfully processed. So the next daily task run would try again (that is, it'd be asked to scan for timestamps over a 48-hour time period instead of the usual 24-hour one).

But this exception handler will cause the run to look successful, so there'll be no retry.

Like I said, though, I'm not actually sure which way is better. If it were just creating notifications in our database, we could have logic to prevent duplicates, so letting the exception bubble up and retrying the next day would be better. But with email, doing that would run the risk of spamming people with redundant notifications if we'd already successfully gotten partway through the list.

File: src/main/resources/templates/email/accessionsAwaitingProcessing/body.ftlh.mjml (hunk 3 above)
Comment (sgrimm) [reply src/main/resources/templates/email/accessionsAwaitingProcessing/body.ftlh.mjml:11]: Ah, that's useful to know for the future. I bet that was a fun one to figure out.

File: src/test/kotlin/com/terraformation/backend/customer/AppNotificationServiceTest.kt [hunk 6]
```diff
@@ -446,4 +458,105 @@ internal class AppNotificationServiceTest : DatabaseTest(), RunsAsUser {
...
+  fun `should store accessions awaiting processing notification`() {
+    // add a second user to check for multiple notifications
+    insertUser(otherUserId)
+    insertOrganizationUser(user.userId, organizationId, Role.CONTRIBUTOR)
+    insertProjectUser(user.userId, projectId, user.userId)
+
+    assertEquals(
+        0,
+        notificationsDao.count(),
+        "There should be no notifications before any notification event")
```
Comment (sgrimm) [inline src/test/kotlin/com/terraformation/backend/customer/AppNotificationServiceTest.kt]: It occurs to me that this is probably not necessary; if the test database is in the wrong initial state, we'll get tons of failures all over the place.

But if it _were_ necessary here, it might be better to do `assertEquals(emptyList<NotificationsRow>(), notificationsDao.findAll(), "...")` so you'd see what the offending data was.

File: src/main/kotlin/com/terraformation/backend/email/EmailNotificationService.kt (hunk 1 above)
Comment (karthikbtf) [reply src/main/kotlin/com/terraformation/backend/email/EmailNotificationService.kt]: Got it, I will revert this change.

File: src/test/kotlin/com/terraformation/backend/customer/AppNotificationServiceTest.kt (hunk 6 above)
Comment (karthikbtf) [reply src/test/kotlin/com/terraformation/backend/customer/AppNotificationServiceTest.kt]: I will remove this check since the table is cleared before every test anyways.

File: src/main/kotlin/com/terraformation/backend/seedbank/daily/StateSummaryNotificationTask.kt (hunk 5 above)
Comment (karthikbtf) [reply src/main/kotlin/com/terraformation/backend/seedbank/daily/StateSummaryNotificationTask.kt]: I see what you mean. `DateNotificationTask` follows a similar pattern.
Looks like notifications are not rolled back either, in the current implementation, that could be fixed.
We could potentially queue up email requests and send them out at the end of the task run (if there are no errors) - spikes in SES sends are usually flagged (would need to throttle the sends), also not sure how we should handle email errors in this case (don't bubble up to avoid spamming users that already had emails sent).
Perhaps a per-user task but that would be overkill?

File: src/main/kotlin/com/terraformation/backend/email/EmailNotificationService.kt [hunk 7]
```diff
@@ -226,10 +239,50 @@ class EmailNotificationService(
...
+              AccessionsFinishedDrying(
+                  config, event.numAccessions, organization.name!!, accessionsUrl)))
+    }
+  }
+
+  @EventListener
+  fun on(@Suppress("UNUSED_PARAMETER") event: DailyTaskTimeArrivedEvent) {
+    accessionDatePendingEmails.clear()
+    accessionStatePendingEmails.clear()
+  }
```
Comment (sgrimm) [inline src/main/kotlin/com/terraformation/backend/email/EmailNotificationService.kt]: I think this makes the intended order of execution a little fuzzy. Event listeners are called synchronously and by default they're called in an undefined order. This could be called before the notifications are generated or after all the processing is done.

Maybe it would make sense to add `DateNotificationTask.StartedEvent` (and similarly for the state summaries) and publish it before doing any real work. Then the code here could clear the pending list based on that event rather than `DailyTaskTimeArrivedEvent`.

File: src/main/kotlin/com/terraformation/backend/email/EmailNotificationService.kt [hunk 8]
```diff
@@ -253,4 +306,11 @@ class EmailNotificationService(
       userStore.fetchByEmail(it)
     }
   }
+
+  companion object data
```
Comment (sgrimm) [inline src/main/kotlin/com/terraformation/backend/email/EmailNotificationService.kt]: Don't need `companion object` here. I assume the intent here was just to define a data class, but FYI, this will end up creating a companion object named `data` that has no methods or properties, and then defining a (non-data) class called `EmailRequest`.

File: src/main/kotlin/com/terraformation/backend/email/EmailNotificationService.kt (hunk 7 above)
Comment (karthikbtf) [reply src/main/kotlin/com/terraformation/backend/email/EmailNotificationService.kt]: I was going to do that,, I should have done that.. thanks for confirming this approach. will do.

File: src/main/kotlin/com/terraformation/backend/email/EmailNotificationService.kt (hunk 8 above)
Comment (karthikbtf) [reply src/main/kotlin/com/terraformation/backend/email/EmailNotificationService.kt]: oops, not what i wanted w.r.t. an empty object named `data`. Thanks for this info.

File: src/main/kotlin/com/terraformation/backend/seedbank/daily/DateNotificationTask.kt [hunk 9]
```diff
@@ -86,11 +87,14 @@ class DateNotificationTask(
     accessionNotificationStore.insertDateNotification(accessionId, message)
   }
 
+  /** Published when the period processed task begins */
+  class PeriodProcessedStartEvent
```
Comment (sgrimm) [inline src/main/kotlin/com/terraformation/backend/seedbank/daily/DateNotificationTask.kt]: These names feel a little awkward to me. Maybe `StartedEvent` and `SucceededEvent`?

Comment (karthikbtf) [reply src/main/kotlin/com/terraformation/backend/seedbank/daily/DateNotificationTask.kt]: `StartedEvent` and `SucceededEvent` sound better/cleaner to me. Thanks for the suggestion.
